### ISSUE TITLE
Provisioning Device Client Registration Result has Exceptions

### ISSUE THREAD

**Issue by haohaowasky** [Azure/azure-iot-sdk-java#432]
- **OS and version used:** Android 4.4

- **Java runtime used:** 1.8 

- **SDK version used:** 1.15 

# Description of the issue:

After 1.15 version SDK released, couple days ago. I was following this [sample code](https://github.com/Azure/azure-iot-sdk-java/blob/master/provisioning/provisioning-samples/provisioning-symmetrickey-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/ProvisioningSymmetricKeySample.java) to call the ProvisioningDeviceClientRegistratinResult function.

In the example code, the SYMMETRIC_KEY String is not very clearly described, I'm not sure what and how to get this symmetric key. 

from the example [here](https://docs.microsoft.com/en-us/azure/iot-dps/how-to-legacy-device-symm-key), I know the logic is something like: 

keybytes=$(echo $KEY | base64 --decode | xxd -p -u -c 1000)
echo -n $REG_ID | openssl sha256 -mac HMAC -macopt hexkey:$keybytes -binary | base64

# Code sample exhibiting the issue:

so in my Android code, here is the logic:

`        byte[] provisioningKey = Base64.decode(BuildConfig.IoTProvisioningKey, Base64.DEFAULT);
        byte[] symmetricKey;

        try {
            // Generate the derived symmetric key based on the provisioning secret and the registration ID
            SecretKeySpec secretKey = new SecretKeySpec(provisioningKey, HMAC_SHA_256);
            Mac hMacSha256 = Mac.getInstance(HMAC_SHA_256);
            hMacSha256.init(secretKey);
            symmetricKey = Base64.encode(hMacSha256.doFinal(registrationId.getBytes()), Base64.NO_WRAP);`

        try {
            provisioningDeviceClient = ProvisioningDeviceClient.create(
                    BuildConfig.IoTGlobalDeviceEndpoint,
                    BuildConfig.IoTScopeId,
                    ProvisioningDeviceClientTransportProtocol.HTTPS,
                    securityClientSymmetricKey);
            ProvisioningStatus provisioningStatus = new ProvisioningStatus();
            provisioningDeviceClient.registerDevice(new ProvisioningDeviceClientRegistrationCallback(), provisioningStatus);

            while (provisioningStatus.provisioningDeviceClientRegistrationInfoClient.getProvisioningDeviceClientStatus() != ProvisioningDeviceClientStatus.PROVISIONING_DEVICE_STATUS_ASSIGNED) {
                ProvisioningDeviceClientStatus status = provisioningStatus.provisioningDeviceClientRegistrationInfoClient.getProvisioningDeviceClientStatus();
                if (status == ProvisioningDeviceClientStatus.PROVISIONING_DEVICE_STATUS_ERROR ||
                        status == ProvisioningDeviceClientStatus.PROVISIONING_DEVICE_STATUS_DISABLED ||
                        status == ProvisioningDeviceClientStatus.PROVISIONING_DEVICE_STATUS_FAILED) {
                    provisioningStatus.exception.printStackTrace();
                    Logger.i("Registration error, bailing out");
                    break;
                }`

I was able to get call back from the function but with an exception complaining about the key is not right.(denied by peers / ssl exceptions / length of the key is over the limit etc)

I feel like it should be an easier way to get the symmetricKey. Please help.

**Comment by haohaowasky** [comment]
> Can I assume that BuildConfig.IoTProvisioningKey points to the Symmetric key of the enrollment itself?
> 
> Can you verify if your call to `registrationId.getBytes()` is using UTF-8?
> 
> We haven't written the utility to handle computing the derived key for Java yet, but we do have some C# code that may be useful to show how it is done. For reference:
> 
> ```cs
> public static string ComputeDerivedSymmetricKey(byte[] masterKey, string registrationId)
> {
>     using (var hmac = new HMACSHA256(masterKey))
>     {
>         return Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(registrationId)));
>     }
> }
> ```

Isn't the computing the derived key has to be done inside of the android app?  The process I assume is we randomly generate a UUID and use that with the provisional key to generate the Symmetric key?

**Comment by timtay-microsoft** [comment]
The computing of the derived symmetric key must be done before turning on your device. Your individual device shouldn't have access to the master level symmetric key.

See [this link](https://docs.microsoft.com/en-us/azure/iot-dps/concepts-symmetric-key-attestation#installation-of-the-derived-device-key) for a more detailed walkthrough

**Comment by haohaowasky** [comment]
> The computing of the derived symmetric key must be done before turning on your device. Your individual device shouldn't have access to the master level symmetric key.
> 
> See [this link](https://docs.microsoft.com/en-us/azure/iot-dps/concepts-symmetric-key-attestation#installation-of-the-derived-device-key) for a more detailed walkthrough

Thanks for the information, 

So what I'm doing here is:

1. turning on the device(first time)
2. the device randomly generate a device ID
3. and call the ProvisioningDeviceClientRegistrationCallback function to register a new device automatically. 

So originally I thought I need a symmetric key, but how do I derive it before the device is turned on?

**Comment by haohaowasky** [comment]
> Can I assume that BuildConfig.IoTProvisioningKey points to the Symmetric key of the enrollment itself?

BuildConfig.IoTProvisioningKey points to the string that I found from Manage enrollment. when I click my name of the IoT hub, it shows Symmetric Keys with one primary and one secondary. 

and then I put them into this function:

        SecurityProviderSymmetricKey securityClientSymmetricKey = new SecurityProviderSymmetricKey(symmetricKey, registrationId);


Please help me out if there is anything wrong...

**Comment by haohaowasky** [comment]
Problem solved, closing it

**Comment by timtay-microsoft** [comment]
Glad to hear you got it figured out. I'm going to close this issue for now. Feel free to open a new issue if you still need any help here.

**Comment by az-iot-builder-01** [comment]
@haohaowasky, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
